### ISSUE TITLE
/etc/localtime is not a symlink

### ISSUE THREAD

**Issue by qis** [Microsoft/BashOnWindows#1392]
As seen [here](https://github.com/Microsoft/BashOnWindows/issues/856) and [here](https://github.com/HowardHinnant/date/issues/102), some applications and libraries rely on `/etc/localtime` to be a symlink.

Manually replacing the file with a symlink to the correct timezone file seems to be a valid workaround that persists through a reboot.

```sh
rm /etc/localtime
ln -s /usr/share/zoneinfo/Europe/Berlin /etc/localtime
```

Executing `dpkg-reconfigure tzdata` breaks it again. Are there any better workarounds? Is this considered a bug and will it be fixed?

**Comment by aseering** [comment]
Hi @qis -- if `dpkg-reconfigure` breaks this on WSL, does it also break this on regular Ubuntu?  WSL is using Ubuntu's binary directly.

**Comment by qis** [comment]
It works on Ubuntu. I only have 16.04.1 LTS for comparison but pretty sure it's the same on 14.01 LTS.

```sh
qis@linux ~ ls -l /etc/localtime
lrwxrwxrwx 1 root root 33 Nov 20 03:10 /etc/localtime -> /usr/share/zoneinfo/Europe/Berlin
qis@linux ~ sudo dpkg-reconfigure tzdata

Current default time zone: 'Europe/Brussels'
Local time is now:      Sun Nov 20 03:10:42 CET 2016.
Universal Time is now:  Sun Nov 20 02:10:42 UTC 2016.

qis@linux ~ ls -l /etc/localtime
lrwxrwxrwx 1 root root 35 Nov 20 03:10 /etc/localtime -> /usr/share/zoneinfo/Europe/Brussels
```

**Comment by therealkenc** [comment]
@aseering - @qis is right. `/etc/localtime` is one of the few things WSL seems to stomp on. So it isn't Real Ubuntu™ in this respect. This is presumably because when you normally install Linux (or whatever dist), one of the few critical things set up is `/etc/localtime` (and `/etc/resolv.conf`) to make glibc happy. What you'll see in a lot of `strace` outputs on WSL is an EINVAL on readlink("/etc/localtime"), which is a horrible pattern that is [unfortunately common](https://www.google.com/search?q=readlink%20%2Fetc%2Flocaltime&rct=j).

Couldn't repro "`dpkg-reconfigure tzdata` breaks it again" though. Now I'm on Berlin time, as far as I can tell.

**Comment by qis** [comment]
Right now there is no way to call either Win32 functions from Kernel32.dll or Win32 system calls.

The `<ctime>` functions report "CET" as the timezone name. This is ambiguous, right?

```cpp
#include <iostream>
#include <ctime>

int main() {
  char buf[16];
  std::time_t time = std::time(nullptr);
  std::strftime(buf, sizeof(buf), "%Z", std::localtime(&time));
  std::cout << buf << std::endl;
}
```

The ICU library also requires `/etc/localtime` to be a symlink and reports "STD" as the timezone name (fallback).

**Comment by therealkenc** [comment]
Yes, ICU is broken in this respect. But after you so the symlink yourself what clobbers it again?

**Comment by qis** [comment]
@therealkenc Nothing. It just requires manual setup once. Not sure if `apt-get upgrade` can break that again. I give up and will wait for Microsoft to resolve this issue. No big deal.

**Comment by therealkenc** [comment]
You said: "Executing `dpkg-reconfigure tzdata` breaks it again". If that was incorrect maybe ~~strikethrough~~ that so people don't get confused.

`apt-get upgrade` won't break anything because `apt-get upgrade` is from Ubuntu not Microsoft, and Ubuntu wants it to be a symlink. [Granted, WSL does have the misguided `lxrun /upgrade` and god knows what that does.]

**Comment by qis** [comment]
@therealkenc Everything I said should still be correct. Executing `dpkg-reconfigure tzdata` replaces a manually created symlink with a file, thus breaking some libraries.

**Comment by therealkenc** [comment]
Interesting. I just did `dpkg-reconfigure tzdata` for a second time to double check, and it updated the symlink, as opposed to replacing it with a file. This is on Xenial on WSL. So it sounds like general Linux borkage and there isn't much to wait for Microsoft on here. As @aseering was saying, that dpkg-reconfigure doesn't play well with ICU is not in Microsoft's wheelhouse.

**Comment by qis** [comment]
@therealkenc Thanks for the info. I'll upgrade as soon as I can find some time.

**Comment by jverne** [comment]
FWIW, I had a fixed `/etc/localtime` file created with the Anniversary version, Ubuntu 14, which persisted through an Insider update to 14965 and subsequent upgrade to Xenial.

Running `dpgk-reconfigure tzdata` replaced it with a symlink and now `date` shows EST instead of STD.

**Comment by NoSubstitute** [comment]
@jverne minor typo, 'dpgk', instead of the correct 'dpkg'.
Yes, I just tried this and it fixes the broken time files.

**Comment by andyneff** [comment]
A little extra information

- I gather that `/etc/localtime` is a hard link to `/usr/share/zoneinfo/Msft/localtime`
- I'm guess that this file is generated by wslhost (specifically code in `C:\Windows\System32\lxss\LxssManager.dll`)

So when you `rm /etc/localtime`, you are breaking that link, and when your host windows timezone change, the WSL no longer knows about that.

1. If all you need is a symlink, instead of determining the timezone to symlink in, you can simply run `sudo ln -sf /usr/share/zoneinfo/Msft/localtime /etc/localtime`
2. I tested on both Ubuntu 14 and 18, and `dpkg-reconfigure tzdata` is still leaving the file a symlink

However, if you need a [non-broken timezone name](https://stackoverflow.com/questions/46352851/what-timezone-does-dst-refer-to), you'll have to follow the many solutions already put forth above.

**Comment by qis** [comment]
The problem I have is that [date](https://howardhinnant.github.io/date/date.html) relies on the path name of the symlink target and not its contents. Since `Msft/localtime` is a non-standard and meaningless timezone *name*, it would be nice to restore the normal Linux behavior.

**Comment by andyneff** [comment]
TL;DR Agree, but in reality, I think the best we can hope for is to have `/etc/localtime` be a symlink by default, rather than a hard link.

---

@qis, I totally agree, it would be nice if it's pointed to a standard meaningful timezone. However the difficulty of [determining a Linux timezone from Windows](https://stackoverflow.com/a/4991854) is not straight forward, future proof, or language proof. So I understand *why* they went with the fake `Msft/localtime` file instead.

I also think the WSL should symlink `/etc/localtime` rather than hard link by default. This would at least *allow* situations like #856 and HowardHinnant/date#102 to successfully run the symlink resolution.  They shouldn't realize that `/usr/share/zoneinfo/Europe/Berlin` is any less nonsense than `/usr/share/zoneinfo/Msft/localtime`

---

*Another* possibility is to set the `TZ` environment value in your `.bashrc` as mentioned [here](https://stackoverflow.com/a/46353158/4166604). For programs that honor `TZ`, this should work. For example, add:

```bash
TZ=/usr/share/zoneinfo/Europe/Berlin
```

And now it doesn't matter who messes up `/etc/localtime`

**Comment by therealkenc** [comment]
> I think the best we can hope for is to have /etc/localtime be a symlink by default, rather than a hard link.

Good post. Someone motivated might fork a new issue making that the explicit ask. Code that depends on it being a symlink is arguably buggy, but using a hardlink when a softlink is the convention is gratuitously different with no redeeming qualities I can think of off hand. [I searched and searched for a LSB case to be made for a symlink, but I didn't succeed.] Maybe the new issue gets marked by-design for reasons or maybe it gets ignored (and that's okay), but I can say with a relative degree of certainty it'll die down here in an unlabeled closed issue from 2016.

**Comment by andyneff** [comment]
@therealkenc Thanks. And I agree: I found #3109 which it's a non-closed issue bringing up just the question about symlinking it, *and* has two Microsoft employees on the thread, 🤞

**Comment by therealkenc** [comment]
Bonus, thanks.
